변 기록(StoolLog) 화면 구현

junu0516 opened the pull request:
### 🔖 관련 이슈
- #14 

<br> 

### ⚒️ 작업 내역

- [x] BottomSheet 컴포넌트 화면에 네비게이션 컨트롤러 적용 확인
- [x] 디자인 시스템 참고해서 화면 구현 완성
- [x] mock 데이터 구성해서 기능 구현
<br>

### 💻 리뷰어 가이드

- 메인홈 화면에서 변 기록하기 버튼 터치 > 만족도 선택 > 세부사항 선택 > 완료버튼 터치 의 순서로 동작합니다.
    - 세부사항은 색, 강도, 크기 항목 모두 선택해야 하며, 선택/비선택 여부에 따라 이미지가 달리 표시됩니다.
    - 단, 크기 항목의 경우에는 선택할 때의 이미지 파일이 없어서 이 부분은 선택/비선택 여부에 관계없이 비선택 이미지로 나타나게 처리한 상태입니다.  
- 팝업 화면을 아래로 쭉 내리거나, 팝업 화면 뒤의 검정 배경을 터치하면 자동으로 팝업화면이 닫겨야 합니다.
- 완료 버튼이 터치되면 팝업 화면이 닫기고, 콘솔창에 선택한 내용이 출력되야 합니다.
- 세부사항 3가지 중 어느 하나라도 선택되지 않으면 완료 버튼은 터치되지 않습니다.
    - 단, 완료 버튼 비활성화에 대한 별도 디자인 표시는 현재 없습니다. 

<br>

### 📝 부가 설명

- 내용도 좀 많고, 기존의 DefaultHomeCoordinator에 팝업 화면 띄우기 위해 임의로 임시 코드를 작성했는데, 한 번 보시고 의도하신대로 다시 고치면 될듯합니다.
- 여기에 다 쓰기는 너무 글이 길어질듯해서.. 컨벤션이나 모듈 배치 등 맞게 했는지 조만간 회의할 때 화상으로 공유하면서 같이 확인하는게 나을듯합니다.
- 혹시 기타 코드 이상하거나 컨벤션 안맞는 부분 코멘트 달아주시면 여기 PR에 수정 커밋 푸시할께요.
<br> 

### 📑 첨부 자료
![Simulator Screen Recording - iPhone 12 mini - 2023-05-07 at 03 14 43](https://user-images.githubusercontent.com/68586291/236640376-a4d85c41-ff98-4707-bf4d-5dcd990b6371.gif)

junu0516 committed:
Merge branch 'develop' into 14-stoolLogScene

junu0516 committed:
#14 feat: 기록 화면에 사용할 image asset 추가

junu0516 committed:
#14 feat: 기록 화면에 사용할 Entity 정의

- CoreEntity 모듈에 있어도 되는지 재확인해야함

junu0516 committed:
#14 feat: FeatureStoolLog 모듈 추가 및 설정

junu0516 committed:
#14 feat: FeatureStoolLog 모듈 추가 관련 누락된 커밋 보충

junu0516 committed:
#14 feat: 디자인 가이드 맞춰서 BottomSheet 변경, BottomSheetController 내부 기능 개선

- BottomSheetController의 서브 뷰를 모두 제거하고, 뷰컨트롤러를 제거하도록 순서 변경
- closeBottomSheet()을 외부에서도 호출 가능하도록 public으로 접근제어 수정

junu0516 committed:
#14 feat: StoolLogCoordinator 및 관련된 필요타입 구현

junu0516 committed:
#14 feat: 기록 화면에 사용할 커스텀 셀 구현

- 색상, 강도, 크기 선택에 대해 각각 ColorSelectionCell, StiffnessSelectionCell, SizeSelectionCell 구현

junu0516 committed:
#14 feat: 만족도 선택을 위한 SatisfactionSelectViewController, SatisfactionSelectViewModel 구현

junu0516 committed:
#14 feat: 세부사항 선택을 위한 SatisfactionDetailViewController, SatisfactionDetailViewModel 구현

junu0516 committed:
#14 feat: 홈화면에서 기록하기 버튼을 통해 기록 화면(BottomSheetController)으로 진입하는 흐름 임시로 구현

junu0516 committed:
Merge branch 'develop' into 14-stoolLogScene

sanghyeok-kim commented on Projects/Features/FeatureStoolLog/FeatureStoolLogPresentation/Sources/Views/SatisfactionDetail/DetailSelection/ColorSelectionCell.swift:
```diff
@@ -0,0 +1,73 @@
+
+import SnapKit
+
+import CoreEntity
+
+import DesignSystem
+
+import Utils
```
내부 모듈끼리는 줄 바꿈 없이 import 해도 됩니다

junu0516 replied:
ㅋㅋ 붙이는걸로 통일해야겠어요

sanghyeok-kim commented on Projects/Features/FeatureStoolLog/FeatureStoolLogPresentation/Sources/Views/SatisfactionSelect/SatisfactionSelectViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        
+        output.disatisfactionButtonSelected
+            .asDriver()
+            .drive(onNext: { [weak self] isDisatisfactionSelected in
+                guard let self = self else { return }
+                
+                let image = isDisatisfactionSelected ?
+                    ImageAsset.badSelected.image :
+                    ImageAsset.badDeselected.image
+                self.disatisfactionButton.setImage(image, for: .normal)
+            })
```
이 로직은 `disatisfactionButton`안에 `setImage(by isSelected: Bool)` 같은 함수를 만들어서
그 안에 구현해도 좋을 것 같아요

그러면 뷰컨트롤러에선 `drive(onNext: disatisfactionButton.setImage)`
이렇게 깔끔하게 쓸 수 있겠네요

junu0516 replied:
@sanghyeok-kim 다시 쭉 보니까 좀 반복되는 패턴들이 있어서.. 이 부분은 코드 개선을 위한 이슈를 별도로 파서 피드백 주신 내용 반영하는 것 포함해서 같이 진행할께요!

sanghyeok-kim commented on Projects/Features/FeatureStoolLog/FeatureStoolLogPresentation/Sources/Views/SatisfactionSelect/SatisfactionSelectViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+                output.disatisfactionButtonSelected.asObservable()
+            )
+            .withUnretained(self)
+            .bind(onNext: { owner, args in
+                let (isSatisfactionSelected, isDisatisfactionSelected) = args
+                guard !isSatisfactionSelected || !isDisatisfactionSelected,
+                      isSatisfactionSelected != isDisatisfactionSelected else { return }
+                
```
`filter`를 활용해도 좋아보여요

junu0516 replied:
@sanghyeok-kim  아 guard문을 아얘 filter로 빼는거 말씀하시는거죠?? 맞으면 그게 훨씬 보기 좋을 것 같아요

sanghyeok-kim commented on Projects/Features/FeatureStoolLog/FeatureStoolLogPresentation/Sources/Views/SatisfactionSelect/SatisfactionSelectViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+    private func bindInputToOutput() {
+        
+        input.didTapSatisfactionButton
+            .withLatestFrom(output.satisfactionButtonSelected)
+            .map { !$0 }
+            .do(onNext: { [weak self] isSatisfactionSelected in
```
`map { !$0 }`을 하신 이유가 있을까요?

sanghyeok-kim replied:
`map { !$0 }` 을 하신 이유가 있을까요?

sanghyeok-kim commented on Projects/Features/FeatureStoolLog/FeatureStoolLogPresentation/Sources/Views/SatisfactionDetail/SatisfactionDetailViewController.swift:
```diff
@@ -0,0 +1,255 @@
+            })
+            .disposed(by: disposeBag)
+        
+        output.selectableColors
+            .bind(to: colorSelectCollectionView.rx.items(
+                cellIdentifier: ColorSelectionCell.identifier,
+                cellType: ColorSelectionCell.self)
+            ) { _, color, cell in
+                cell.configure(selectableColor: color)
+            }
+            .disposed(by: disposeBag)
+        
+        output.selectableStiffnessList
+            .bind(to: stiffnessSelectCollectionView.rx.items(
+                cellIdentifier: StiffnessSelectionCell.identifier,
+                cellType: StiffnessSelectionCell.self)
+            ) { _, stiffness, cell in
+                cell.configure(stiffnessSelection: stiffness)
+            }
+            .disposed(by: disposeBag)
+        
+        output.selectableSizes
+            .bind(to: sizeSelectionCollectionView.rx.items(
```
`asDriver`가 누락되었네요

junu0516 replied:
그러네요! 이건 Driver 사용하는걸로 고쳐놓을께요

하나 좀 궁금해지는게 `bind(to:)` 안에 Rxcocoa의 오퍼레이터가 들어가는데, Rxcocoa면 내부적으로 알아서 메인스레드에서 작동하게 하는지 확인해봐야겠어요!

junu0516 replied:
[커밋](https://github.com/LifePoop/LifePoop_iOS/pull/21/commits/2c24f1b2cb1a2026c65fb77b5554f9d8fe721428) 에 나온대로 우선 고쳐봤어요..

asDriver을 사용하면 drive(onNext:)로 클로저 타입으로 명시해야 하는데.. 코드가 좀 복잡해지는거 같아서 observe(on:MainScheduler.asyncInstance)를 사용하는 식으로 우선 처리해봤습니다.. 혹시 drive(onNext:)를 사용한다고 하면 어떻게 수정하면 될까요??
저는 대충 아래처럼 작성하는 방식만 떠오르네요 🥲
```swift
output.selectableColors
    .asDriver(onErrorJustReturn: [])
    .drive(onNext: { colors in
        Observable.just(colors)
            .bind(to: colorSelectCollectionView.rx.items(
                cellIdentifier: ColorSelectionCell.identifier,
                cellType: ColorSelectionCell.self
            )) { _, color, cell in
                cell.configure(selectableColor: color)
            }
            .disposed(by: disposeBag)
    })
    .disposed(by: disposeBag)
```

junu0516 committed:
#14 refactor: import 구문 정렬 수정

junu0516 committed:
#14 refactor: 만족/불만족 여부 관련 바인딩에 filter 오퍼레이터 적용 추가

junu0516 committed:
#14 refactor: 세부사항 선택 화면에서, 컬렉션뷰 업데이트 할 때 observe(on:) 오퍼레이터 추가

junu0516 committed:
#14 refactor: 만족/불만족 버튼 선택할 때마다 이미지 바꾸는 로직 수정

- 아얘 .normal, .selected 일 때 다른 이미지가 할당되도록 초기화 시점에 결정하도록 수정
- 이후에는 각 버튼의 rx.isSelected 속성에 Bool 값을 방출시켜서 버튼 이미지가 자동으로 바뀌도록 처리

junu0516 committed:
#14 feat: UIView+roundCorners 개선

- 모든 모서리가 아닌, 원하는 부분만 선택해서 적용하도록 수정
- radius 파라미터명 명시하도록 수정

junu0516 committed:
#14 feat: BottomSheet 모서리 둥글게 처리하는 부분 상단에만 적용되도록 수정
